---
title: ""
output: github_document
---

```{r eval=FALSE, echo=FALSE, message=FALSE}
# Use this command to render the file
rmarkdown::render('Table PR2 Primers.Rmd', 
                  output_file = "C:/daniel.vaulot@gmail.com/Databases/_PR2/primers/wiki/18S-rRNA-primers.md", 
                  output_format="github_document", 
                  encoding = 'UTF-8')
```

A compilation of 18S rRNA primers from the literature, mapped onto the reference SSU sequence for _[Saccharomyces cerevisiae](http://apollo.chemistry.gatech.edu/RibosomeGallery/S%20cerevisiae/SSU/3D%20structure%20based/index.html#S_cerevisiae.png)_ ([FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071), 1799 nucleotides, first nucleotide marked as 1) .

**Trick**: If you are using Chrome, use this [extension](https://chrome.google.com/webstore/detail/wide-github/kaalofacklcidaampbokdplbklpeldpj) to view the full width of the table.

Columns are as follows

 column               | definition 
| :------------------ | :------- |
 primer_id              | primer id in pr2-primers database
 name                | original name of the primer 
 synonyms              | synonyms found in the litterature 
 sequence  |
 sequence revcomp | reverse complement of the sequence
 direction | forward (fwd) or reverse (rev)
 start | start of primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
 end | end of primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
 specificity | is the primer specific of a group
 reference | original reference where primer was first defined
 doi | link to original paper
 

Please report errors or primers not listed here in the [Issues page of the PR2 primers database](https://github.com/pr2database/pr2-primers/issues).

#### Contributors
* Stefan Geisen: S.Geisen@nioo.knaw.nl   
* Fréderic Mahé: frederic.mahe@cirad.fr   
* Daniel Vaulot: vaulot@gmail.com
* Enrique Lara: enrique.lara@rjb.csic.es   



```{r init, eval=TRUE, message=FALSE, warning=FALSE, echo=FALSE}

# Libraries tidyr ---------------------------------------------------------

  library("ggplot2") 
  library("dplyr")   
  library("readxl")
  library("stringr")
  library("Biostrings")
  library("dvutils")

# Options for knitting the report -------------  
  
  library(knitr)
  library(rmdformats)
  library(kableExtra) 

  knitr::opts_chunk$set(fig.width=8, 
                        fig.height=6, 
                        eval=TRUE, 
                        cache=TRUE,
                        echo=TRUE,
                        prompt=FALSE,
                        tidy=TRUE,
                        comment=NA,
                        message=FALSE,
                        warning=FALSE)
  opts_knit$set(width=90)
  options(max.print="500")  
  options(knitr.kable.NA = '')
```



```{r, echo=FALSE, message=FALSE, warning=FALSE}
# primers <- read_excel(path = "primers_18S.xlsx", sheet ="primers")

  pr2_db <- db_info("pr2_local")
  pr2_db_con <- db_connect(pr2_db)
  primers <- tbl(pr2_db_con, "pr2_primers") %>% collect()
  disconnect <- db_disconnect(pr2_db_con)
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}

sequence_revcomp=data.frame(sequence_revcomp = as.character(reverseComplement(DNAStringSet(primers$sequence))))

primers_wiki <- primers %>% 
  bind_cols(sequence_revcomp) %>% 
  filter(gene=="18S_rRNA") %>% 
  arrange(start_yeast) %>% 
  mutate(doi = str_c("https://doi.org/", doi)) %>% 
  select(primer_id, name, synonyms, sequence, `sequence revcomp`= sequence_revcomp,direction, start=start_yeast, end=end_yeast, specificity, reference, doi)

knitr::kable(primers_wiki, format = 'markdown')



```
