---
title: ""
output: github_document
---

```{r eval=FALSE, echo=FALSE, message=FALSE}
# Use this command to render the file
rmarkdown::render('Table PR2 Primer Sets.Rmd', 
                  output_file = 'C:/daniel.vaulot@gmail.com/Databases/_PR2/primers/wiki/18S-rRNA-primer-sets.md', 
                  output_format="github_document", 
                  encoding = 'UTF-8')
```

A compilation of 18S rRNA primer sets for metabarcoding compiled from the literature, mapped onto the reference SSU sequence for _[Saccharomyces cerevisiae](http://apollo.chemistry.gatech.edu/RibosomeGallery/S%20cerevisiae/SSU/3D%20structure%20based/index.html#S_cerevisiae.png)_ ([FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071), 1799 nucleotides, first nucleotide marked as 1) .

**Trick**: If you are using Chrome, use this [extension](https://chrome.google.com/webstore/detail/wide-github/kaalofacklcidaampbokdplbklpeldpj) to view the full width of the table.

Columns are as follows

 column               | definition 
| :------------------ | :------- |
  region | gene region targetted. Universal corresponds to primer sets that amplify both prokaryotes and eukaryotes
  specificity |  is the primer specific of a group
  primer_set_id | id of the primer set  in pr2-primers database
  primer_set_name |  name of the primer set (usually from author)
  fwd_name | forward primer name
  fwd_seq | forward primer sequence
  fwd_start | start of forward primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
  fwd_end | end of forward primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
  rev_name | reverse primer name
  rev_seq | reverse primer sequence
  rev_start | start of reverse primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
  rev_end | end of reverse primer relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
  length | length of amplified sequence relative to [FU970071](https://www.ncbi.nlm.nih.gov/nuccore/FU970071)
  reference | original reference where primer set was first defined
  doi | link to original paper
 

Please report errors or primer sets not listed here in the [Issues page of the PR2 primer database](https://github.com/pr2database/pr2-primers/issues).

#### Contributors
* Stefan Geisen: S.Geisen@nioo.knaw.nl   
* Fréderic Mahé: frederic.mahe@cirad.fr   
* Daniel Vaulot: vaulot@gmail.com
* Enrique Lara: enrique.lara@rjb.csic.es



```{r init, eval=TRUE, message=FALSE, warning=FALSE, echo=FALSE}

# Libraries tidyr ---------------------------------------------------------

  library("ggplot2") 
  library("dplyr")   
  library("readxl")
  library("stringr")
  library("Biostrings")

# Options for knitting the report -------------  
  
  library(knitr)
  library(rmdformats)
  library(kableExtra) 

  knitr::opts_chunk$set(fig.width=8, 
                        fig.height=6, 
                        eval=TRUE, 
                        cache=TRUE,
                        echo=TRUE,
                        prompt=FALSE,
                        tidy=TRUE,
                        comment=NA,
                        message=FALSE,
                        warning=FALSE)
  opts_knit$set(width=90)
  options(max.print="500")  
  options(knitr.kable.NA = '')
```



```{r, echo=FALSE, message=FALSE, warning=FALSE}
# primers <- read_excel(path = "primers_18S.xlsx", sheet ="primers")

  pr2_db <- db_info("pr2_local")
  pr2_db_con <- db_connect(pr2_db)
  primers <- tbl(pr2_db_con, "pr2_primers") %>% collect()
  primer_sets_all <- tbl(pr2_db_con, "pr2_primer_sets") %>% collect()
  disconnect <- db_disconnect(pr2_db_con)
  
  # Just keep the 18S rRNA primers
  primer_sets <- primer_sets_all %>% 
      filter(gene ==  "18S_rRNA") 
      
  
  primer_sets <- primer_sets %>% 
    left_join(select(primers, 
                     primer_id, 
                     fwd_name=name,
                     fwd_seq=sequence, 
                     fwd_start_yeast= start_yeast, 
                     fwd_end_yeast= end_yeast), 
              by = c("fwd_id" = "primer_id")) %>% 
    left_join(select(primers, 
                   primer_id, 
                   rev_name=name,
                   rev_seq=sequence, 
                   rev_start_yeast= start_yeast, 
                   rev_end_yeast= end_yeast), 
            by = c("rev_id" = "primer_id")) %>% 
    mutate(length_yeast = rev_end_yeast - fwd_start_yeast + 1) %>% 
    select(primer_set_id, gene_region, specificity, primer_set_id,primer_set_name, contains("fwd"), contains("rev"),length_yeast, used_in:remark) %>% 
    arrange(gene_region, specificity)
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}

primer_sets_wiki <-primer_sets %>% 
  mutate(doi = str_c("https://doi.org/", reference_doi)) %>%
  arrange(fwd_start_yeast) %>%
  rename_all(funs(str_replace(., "_yeast", ""))) %>% 
  select(gene_region, specificity,primer_set_id, primer_set_name,  contains("fwd"), contains("rev"), length, reference, doi, -fwd_id, -rev_id)

knitr::kable(primer_sets_wiki, format = 'markdown')



```
